---
layout: post
title: java.util.ConcurrentModificationException异常解决
---

java.util.ConcurrentModificationException异常解决

在执行下面的代码

    import java.util.ArrayList;
    import java.util.List;

    public class Test {
        public static void main(String[] args){

            List<string> list = new ArrayList</string><string>();
            list.add("b");
            list.add("a");
            list.add("a");
            list.add("c");

            for (String s : list) {
                if ("a".equals(s)) {
                    list.remove(s);
                }
            }
            System.out.println(list.size());
        }
    }


会抛出下面的异常

    Exception in thread "main" java.util.ConcurrentModificationException
        at java.util.AbstractList$Itr.checkForComodification(Unknown Source)
        at java.util.AbstractList$Itr.next(Unknown Source)
        at Test.main(Test.java:13)

    
我们知道加强的for循环内部也是Iterator实现的
Iterator 是工作在一个独立的线程中，并且拥有一个 mutex 锁。 Iterator 被创建之后会建立一个指向原来对象的单链索引表，当原来的对象数量发生变化时，这个索引表的内容不会同步改变，所以当索引指针往后移动的时候就找不到要迭代的对象，所以按照 fail-fast 原则 Iterator 会马上抛出 java.util.ConcurrentModificationException 异常。

　　所以 Iterator 在工作的时候是不允许被迭代的对象被改变的。但你可以使用 Iterator 本身的方法 remove() 来删除对象， Iterator.remove() 方法会在删除当前迭代对象的同时维护索引的一致性。

　　有意思的是如果你的 Collection / Map 对象实际只有一个元素的时候， ConcurrentModificationException 异常并不会被抛出。这也就是为什么在 javadoc 里面指出： it would be wrong to write a program that depended on this exception for its correctness: ConcurrentModificationException should be used only to detect bugs.

所以我们只要把删除字符串是a的循环语句改为如下就可以了

    for(Iterator it=list.iterator();it.hasNext();){
            String s = (String) it.next();
            if ("a".equals(s)) {
                it.remove();
            }
        }
